src/polygon_filter.cpp

```diff
@@ -0,0 +1,412 @@
+  {
+    ROS_WARN("Could not get transform, ignoring laser scan! %s", error_msg.c_str());
+    return false;
+  }
+
+  try
+  {
+    projector_.transformLaserScanToPointCloud(polygon_frame_, input_scan, laser_cloud, tf_);
+  }
+  catch (tf::TransformException& ex)
+  {
+    ROS_INFO_THROTTLE(.3, "Ignoring Scan: Waiting for TF");
+    return false;
+  }
+  const int i_idx_c = sensor_msgs::getPointCloud2FieldIndex(laser_cloud, "index");
+  const int x_idx_c = sensor_msgs::getPointCloud2FieldIndex(laser_cloud, "x");
+  const int y_idx_c = sensor_msgs::getPointCloud2FieldIndex(laser_cloud, "y");
+  const int z_idx_c = sensor_msgs::getPointCloud2FieldIndex(laser_cloud, "z");
+
+  if (i_idx_c == -1 || x_idx_c == -1 || y_idx_c == -1 || z_idx_c == -1)
+  {
+    ROS_INFO_THROTTLE(.3, "x, y, z and index fields are required, skipping scan");
+  }
+
+  const int i_idx_offset = laser_cloud.fields[i_idx_c].offset;
+  const int x_idx_offset = laser_cloud.fields[x_idx_c].offset;
+  const int y_idx_offset = laser_cloud.fields[y_idx_c].offset;
+  const int z_idx_offset = laser_cloud.fields[z_idx_c].offset;
+
+  const int pstep = laser_cloud.point_step;
+  const long int pcount = laser_cloud.width * laser_cloud.height;
+  const long int limit = pstep * pcount;
+
+  int i_idx, x_idx, y_idx, z_idx;
+  for (i_idx = i_idx_offset, x_idx = x_idx_offset, y_idx = y_idx_offset, z_idx = z_idx_offset;
+
+       x_idx < limit;
+
+       i_idx += pstep, x_idx += pstep, y_idx += pstep, z_idx += pstep)
+  {
+    // TODO works only for float data types and with an index field
+    // I'm working on it, see https://github.com/ros/common_msgs/pull/78
+    float x = *((float*)(&laser_cloud.data[x_idx]));
+    float y = *((float*)(&laser_cloud.data[y_idx]));
+    float z = *((float*)(&laser_cloud.data[z_idx]));
+    int index = *((int*)(&laser_cloud.data[i_idx]));
+
+    tf::Point point(x, y, z);
+
+    if (!invert_filter_)
+    {
+      if (inPolygon(point))
+      {
+        output_scan.ranges[index] = std::numeric_limits<float>::quiet_NaN();
+      }
+    }
+    else
+    {
+      if (!inPolygon(point))
+      {
+        output_scan.ranges[index] = std::numeric_limits<float>::quiet_NaN();
+      }
+    }
+  }
+
+  return true;
+}
+
+
+bool LaserScanPolygonFilter::inPolygon(tf::Point& point) const
+{
+  int i, j;
+  bool c = false;
+
+  for (i = 0, j = polygon_.points.size() - 1; i < polygon_.points.size(); j = i++)
+  {
+    if ((polygon_.points.at(i).y > point.y() != (polygon_.points.at(j).y > point.y()) &&
+         (point.x() < (polygon_.points[j].x - polygon_.points[i].x) * (point.y() - polygon_.points[i].y) /
+                              (polygon_.points[j].y - polygon_.points[i].y) +
+                          polygon_.points[i].x)))
+      c = !c;
+  }
+  return c;
+}
+
+
+void LaserScanPolygonFilter::reconfigureCB(laser_filters::PolygonFilterConfig& config, uint32_t level)
+{
+  invert_filter_ = config.invert;
+  polygon_ = makePolygonFromString(config.polygon);
```

Comment:
It looks like `makePolygonFromString()` will return an empty polygon if the string isn't properly formatted. Could that cause problems? If `invert` is set to true, would an empty polygon filter out all laser points?

Reply:
Yes, this is intended. What behavior do you suggest otherwise?

Reply:
I haven't thought this through deeply, but what do you think of:
- on startup, if there is no valid polygon string, die
- if someone accidentally inputs an invalid polygon using dynamic reconfigure, keep the old polygon.

Mostly I'm trying to avoid the case where a obstacle avoidance starts ignoring all laser points.